src/coreclr/tools/Common/Compiler/TypeWithRepeatedFields.cs

```diff
@@ -0,0 +1,170 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using Internal.TypeSystem;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// This type represents a type that has one field in metadata,
+    /// but has that field repeated at runtime to represent an array of elements inline.
+    /// </summary>
+    public sealed partial class TypeWithRepeatedFields : MetadataType
+    {
+        private int? _numFields;
+        private FieldDesc[] _fields;
+
+        public TypeWithRepeatedFields(MetadataType underlyingType)
+        {
+            MetadataType = underlyingType;
+        }
+
+        internal MetadataType MetadataType { get; }
+
+        internal int NumFields
+        {
+            get
+            {
+                if (_numFields.HasValue)
+                {
+                    return _numFields.Value;
+                }
+                int size = MetadataType.InstanceFieldSize.AsInt;
+                FieldDesc firstField = GetFirstInstanceField();
+
+                _numFields = size / firstField.FieldType.GetElementSize().AsInt;
+                return _numFields.Value;
+            }
+        }
+
+        private FieldDesc GetFirstInstanceField()
+        {
+            FieldDesc firstField = null;
+            foreach (var field in MetadataType.GetFields())
+            {
+                if (field.IsStatic)
+                {
+                    continue;
+                }
+
+                firstField = field;
+                break;
+            }
+
+            Debug.Assert(firstField is not null);
+            return firstField;
+        }
+
+        private FieldDesc[] ComputeFields()
+        {
+            var fields = new FieldDesc[NumFields];
+
+            FieldDesc firstField = GetFirstInstanceField();
+
+            for (int i = 0; i < NumFields; i++)
+            {
+                fields[i] = new ImpliedRepeatedFieldDesc(this, firstField, i);
+            }
+
+            return fields;
+        }
+
+        public override IEnumerable<FieldDesc> GetFields()
+        {
+            if (_fields is null)
+            {
+                Interlocked.CompareExchange(ref _fields, ComputeFields(), null);
+            }
+            return _fields;
+        }
+
+        public override ClassLayoutMetadata GetClassLayout() => MetadataType.GetClassLayout();
+        public override bool HasCustomAttribute(string attributeNamespace, string attributeName) => MetadataType.HasCustomAttribute(attributeNamespace, attributeName);
+        public override IEnumerable<MetadataType> GetNestedTypes() => MetadataType.GetNestedTypes();
```

Comment:
We don't dispatch back to the underlying type for things that shouldn't matter, see e.g. https://github.com/dotnet/runtime/blob/b869e8da11161de3102e2e966ce8f0a61d9ccbd4/src/coreclr/tools/Common/TypeSystem/Interop/IL/NativeStructType.cs.

It makes it less likely that we'd accidentally leak this special type into the rest of the system where it would kind of look like the original type.

Things like methods on the type, methodimpls, implemented interfaces, etc. can probably be all empty.

Other things like "IsSealed", "IsAbstract", "has finalizer", "has variance" etc. - we know the answer - it's a struct (and ideally also non-generic struct or there's extra trouble we'd need to go though to represent fields on instantiated types).

Reply:
I'd expect the generic case to be something we might need to care about, as we allow InlineArray on generic types (and we use `TypeWithRepeatedFields` for the SystemV ABI calculation for managed and unmanaged types).